### ISSUE TITLE
Pets crash the server

### ISSUE THREAD

**Issue by DenielWorld** [BlockHorizons/BlockPets#176]
If multiple people have a pet for at least 5 minutes, the server crashes.

**Comment by ericsimard52** [comment]
I think this would more then likely be related to https://github.com/BlockHorizons/BlockPets/issues/147

**Comment by DenielWorld** [comment]
That is resolved now, but now pets crash a player when he kills one.

**Comment by ericsimard52** [comment]
I notice that too. When you reenter the game pet is back full health.

**Comment by DenielWorld** [comment]
I might be able to get a log, but I'd need your help.

**Comment by ericsimard52** [comment]
What do you need?

**Comment by ericsimard52** [comment]
It does not cause a crash dump with me, but it crashes before the pet die with the following backtrace:

> [Server thread/CRITICAL]: Error: "Call to a member function getServer() on null" (EXCEPTION) in "src/pocketmine/event/entity/EntityDamageByEntityEvent" at line 72
[19:10:00] [Server thread/DEBUG]: #0 src/pocketmine/entity/Living(568): pocketmine\event\entity\EntityDamageByEntityEvent->getDamager()
[19:10:00] [Server thread/DEBUG]: #1 BlockPets/src/BlockHorizons/BlockPets/pets/BasePet(337): pocketmine\entity\Living->attack(pocketmine\event\entity\EntityDamageByEntityEvent object)
[19:10:00] [Server thread/DEBUG]: #2 BlockPets/src/BlockHorizons/BlockPets/pets/IrasciblePet(72): BlockHorizons\BlockPets\pets\BasePet->attack(pocketmine\event\entity\EntityDamageByEntityEvent object)
[19:10:00] [Server thread/DEBUG]: #3 src/pocketmine/Player(2521): BlockHorizons\BlockPets\pets\IrasciblePet->attack(pocketmine\event\entity\EntityDamageByEntityEvent object)
[19:10:00] [Server thread/DEBUG]: #4 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(139): pocketmine\Player->handleInventoryTransaction(pocketmine\network\mcpe\protocol\InventoryTransactionPacket object)
[19:10:00] [Server thread/DEBUG]: #5 src/pocketmine/network/mcpe/protocol/InventoryTransactionPacket(155): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleInventoryTransaction(pocketmine\network\mcpe\protocol\InventoryTransactionPacket object)
[19:10:00] [Server thread/DEBUG]: #6 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(99): pocketmine\network\mcpe\protocol\InventoryTransactionPacket->handle(pocketmine\network\mcpe\PlayerNetworkSessionAdapter object)
[19:10:00] [Server thread/DEBUG]: #7 src/pocketmine/network/mcpe/protocol/BatchPacket(114): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleDataPacket(pocketmine\network\mcpe\protocol\InventoryTransactionPacket object)
[19:10:00] [Server thread/DEBUG]: #8 src/pocketmine/network/mcpe/PlayerNetworkSessionAdapter(99): pocketmine\network\mcpe\protocol\BatchPacket->handle(pocketmine\network\mcpe\PlayerNetworkSessionAdapter object)
[19:10:00] [Server thread/DEBUG]: #9 src/pocketmine/Player(3061): pocketmine\network\mcpe\PlayerNetworkSessionAdapter->handleDataPacket(pocketmine\network\mcpe\protocol\BatchPacket object)
[19:10:00] [Server thread/DEBUG]: #10 src/pocketmine/network/mcpe/RakLibInterface(162): pocketmine\Player->handleDataPacket(pocketmine\network\mcpe\protocol\BatchPacket object)
[19:10:00] [Server thread/DEBUG]: #11 vendor/pocketmine/raklib/src/server/ServerHandler(98): pocketmine\network\mcpe\RakLibInterface->handleEncapsulated(string 192.168.0.10 49029, raklib\protocol\EncapsulatedPacket object, integer 0)
[19:10:00] [Server thread/DEBUG]: #12 src/pocketmine/network/mcpe/RakLibInterface(102): raklib\server\ServerHandler->handlePacket()
[19:10:00] [Server thread/DEBUG]: #13 src/pocketmine/network/mcpe/RakLibInterface(92): pocketmine\network\mcpe\RakLibInterface->process()
[19:10:00] [Server thread/DEBUG]: #14 vendor/pocketmine/snooze/src/SleeperHandler(120): pocketmine\network\mcpe\RakLibInterface->pocketmine\network\mcpe\{closure}()
[19:10:00] [Server thread/DEBUG]: #15 vendor/pocketmine/snooze/src/SleeperHandler(82): pocketmine\snooze\SleeperHandler->processNotifications()
[19:10:00] [Server thread/DEBUG]: #16 src/pocketmine/Server(2278): pocketmine\snooze\SleeperHandler->sleepUntil(double 1546737000.6685)
[19:10:00] [Server thread/DEBUG]: #17 src/pocketmine/Server(2139): pocketmine\Server->tickProcessor()
[19:10:00] [Server thread/DEBUG]: #18 src/pocketmine/Server(1715): pocketmine\Server->start()
[19:10:00] [Server thread/DEBUG]: #19 src/pocketmine/PocketMine(249): pocketmine\Server->__construct(BaseClassLoader object, pocketmine\utils\MainLogger object, string /home/papa/PMMP/, string /home/papa/PMMP/plugins/)
[19:10:00] [Server thread/DEBUG]: #20 /home/papa/PMMP/PocketMine-MP.phar(1): require(string phar:///home/papa/PMMP/PocketMine-MP.phar/src/pocketmine/PocketMine.php)

**Comment by DenielWorld** [comment]
I need someone on my server to kill a pet to see the crashdump

**Comment by ericsimard52** [comment]
Got the crash dump, the above error happen when a player attempted to kill my pet, I think it crashed before the pet died and only caused an error with no crashdump.
This time, I spawned multiple spider creature and got them to kill my pet. It created a crashdump this time, seems to be the same error.
[error.log](https://github.com/BlockHorizons/BlockPets/files/2730084/error.log)

**Comment by ericsimard52** [comment]
The crash happens after BasePet->attack calls parent::attack($source)
I got the debug to tell me the pet health before calling parent::attack, everytime pet's health was at 2 just before the crash.

**Comment by DenielWorld** [comment]
So is there a possible fix?

**Comment by ericsimard52** [comment]
I'm pretty sure there is, I will ask a little patience though, the 2 regular are a little short on time for this, I just started contributing to it.

**Comment by ericsimard52** [comment]
A work around for now is to edit config.yml under plugin_data/BlockPets and set Invulnerable-Pets: true

**Comment by ericsimard52** [comment]
I have been digging trying to find the exact place where this crash happened, I learned a lot, but I still haven't figured out exactly where it happens.
Although it looks like the server crash when the pets die, there is a lot happening between death and crash.
I attempted to pinpoint the problematic line by using return statement and following function calls
The pets gets kill, the BasePet->onDeath executes.
onDeath will not cause a crash until the $newPet = $loader->clonePet($this); execute.
Inside clonePet, we will call loader->createPet.
Inside of this function if I return $entity before $this->playerPets[$player->getLower....] the server does not crash.
It does not make much sense to me, hopefully it will to someone else.
If someone has any pointer I will do the work.

**Comment by ericsimard52** [comment]
When returning entity before updating playerPets array does not cause a server crash, but also does not respawn the pet.
Attempt to toggle the pet afterward results with a crash with the following error:

> [Server thread/CRITICAL]: TypeError: "Argument 1 passed to BlockHorizons\BlockPets\commands\BaseCommand::sendWarning() must implement interface pocketmine\command\CommandSender, string given, called in /home/papa/PMMP/plugins/BlockPets/src/BlockHorizons/BlockPets/commands/TogglePetCommand.php on line 92" (EXCEPTION) in "BlockPets/src/BlockHorizons/BlockPets/commands/BaseCommand" at line 43
[Mon_Jan_7-16.19.53-CST_2019.log](https://github.com/BlockHorizons/BlockPets/files/2734485/Mon_Jan_7-16.19.53-CST_2019.log)

**Comment by ericsimard52** [comment]
Ok, the crash happen $player->getLowerCaseName() 
Looks like $player is properly set

**Comment by ericsimard52** [comment]
Got a bit more understanding, there is something weird and problematic going on with  player and pet name when there is capitals.
As mentioned above in Loader->createPet, if I change player->getLowerCaseName, to player->getName pet dies and respawn a few seconds later. But, this break togglePet which need to have player in lowercase.
After playing around with it, I also had the issue come up where the pet would not move when being ridden, changing petOwner name and pet name to lowercase fixed that.
Would we be better to go through the code and not use lowercase name? Any other ideas?

**Comment by ericsimard52** [comment]
There is issue coming from case on petOwner name and petname.
After changing the above in createPet, if fixed the pet dying, but I was unable to togglepet after that.
in loader->getPetByName I have to remove the strtolower and getLowCaseName for toggle to work, but this broke pet death again.
For myself right now I made a second copy of getPetByName to support toggle.
Also after those mods, for some reason my pets would not move when being ridden, I change my petOwner name to lowercase and fixed the later.
So I have achieved stability  to ride, toggle, tp pet and pets death. But not in a acceptable code yet.
Any reason why we have to play this much with name capitalization?
Would it not be better to stay case sensitive? I first thought to lowercase before we write to database, but that can cause an issue when comparing petOwner to Player name when the later has a capital in it.
